core/embed/lib/error_handling.c

```diff
@@ -0,0 +1,116 @@
+/*
+ * This file is part of the Trezor project, https://trezor.io/
+ *
+ * Copyright (c) SatoshiLabs
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <stddef.h>
+
+#include "common.h"
+#include "display.h"
+#include "error_handling.h"
+#include "mini_printf.h"
+#ifdef FANCY_FATAL_ERROR
+#include "rust_ui.h"
+#else
+#include "terminal.h"
+#endif
+
+#ifdef RGB16
+#define COLOR_FATAL_ERROR RGB16(0x7F, 0x00, 0x00)
+#else
+#define COLOR_FATAL_ERROR COLOR_BLACK
+#endif
+
+void __attribute__((noreturn))
+error_shutdown_ex(const char *title, const char *message, const char *footer) {
+  if (title == NULL) {
+    title = "INTERNAL ERROR";
+  }
+  if (footer == NULL) {
+    footer = "PLEASE VISIT\nTREZOR.IO/RSOD";
+  }
+
+#ifdef FANCY_FATAL_ERROR
+  error_shutdown_rust(title, message, footer);
+#else
+  display_orientation(0);
+  term_set_color(COLOR_WHITE, COLOR_FATAL_ERROR);
+  if (title) {
+    term_printf("%s\n", title);
+  }
+  if (message) {
+    term_printf("%s\n", message);
+  }
+  if (footer) {
+    term_printf("\n%s\n", footer);
+  }
+  display_backlight(255);
+  trezor_shutdown();
+#endif
+}
+
+void __attribute__((noreturn)) error_shutdown(const char *message) {
+  error_shutdown_ex(NULL, message, NULL);
+}
+
+void __attribute__((noreturn))
+__fatal_error(const char *msg, const char *file, int line) {
+#ifdef FANCY_FATAL_ERROR
+  if (msg == NULL) {
+    char buf[128] = {0};
+    mini_snprintf(buf, sizeof(buf), "%s: %d", file, line);
```

Comment:
nit that always bothered me: should there be a space between `file: lineno`? i'm much more used to no space here

Reply:
no problem, I changed it, now it's unified with the similar code in Rust, see d6f4168466119bdce402b112965816644c3cc6e9